apicula.dockerfile

```diff
@@ -0,0 +1,46 @@
+# Authors:
+#   Unai Martinez-Corral
+#   Lucas Teske
+#
+# Copyright 2019-2021 Unai Martinez-Corral <unai.martinezcorral@ehu.eus>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+FROM hdlc/build:build AS build
+
+RUN apt-get update -qq \
+ && DEBIAN_FRONTEND=noninteractive apt-get -y install --no-install-recommends wget
+
+RUN mkdir /opt/apicula \
+ && wget https://files.pythonhosted.org/packages/1a/d6/b3162f87ff114d639095fe7c0655080ee16caff9037d6629f738d8b28d92/Apycula-0.0.1a6.tar.gz \
+ && tar -xvf Apycula-0.0.1a6.tar.gz -C /opt/apicula \
+ && mv /opt/apicula/Apycula-0.0.1a6/* /opt/apicula \
+ && rm -fr /opt/apicula/Apycula-0.0.1a6/ \
+ && ls -lah /opt/apicula
+
+#---
+
+FROM scratch AS pkg
+COPY --from=build /opt/apicula /apicula
+
+#---
+
+FROM hdlc/build:base
+RUN apt-get update -qq \
+ && DEBIAN_FRONTEND=noninteractive apt-get -y install --no-install-recommends python3 python3-setuptools
+
+COPY --from=build /opt/apicula /opt/apicula
+RUN cd /opt/apicula && python3 setup.py install
```

Comment:
This RUN command should not be done here, but in the build stage. Therefore, python3-setuptools should not be required in the final image.

Reply:
Well there is no way to install apycula without that. It is needed for that image otherwise the `gowin_pack` and `gowin_unpack` will not be usable. These commands are aliases to python functions that is created by setup tools.

Reply:
Well, that is undesirable, because it requires a non standard procedure for this tool, which we need to replicate in all the images where apicula is to be used. I would strongly recommend having the build procedure fixed/enhanced upstream, so that it's installable similarly to Symbiyosys.

Anyway, let's move forward. Please, do the following modifications:

- `python3` is already installed in `hdlc/build:base` (see `base.dockerfile`). There is no need to install it here. Furthermore, please install `python3-setuptools` in https://github.com/hdl/containers/blob/main/base.dockerfile#L26. Then, remove apt-get calls from here.
- Instead of using `COPY`, use the experimental mount:
  - Add this comment at the top of the dockerfile: https://github.com/ghdl/ghdl-cosim/blob/master/docker/vunit-cocotb.dockerfile#L1.
  - Modify the `RUN` call similarly to https://github.com/ghdl/ghdl-cosim/blob/master/docker/vunit-cocotb.dockerfile#L34.
- Modify `test/apicula.pkg.sh`, for taking into account that no content will be located in `/apicula/usr/local`, only in `/apicula`.

Reply:
You can actually run the scripts with `python -m apycula.gowin_pack` without installation, the trick is just that the module path has to be set correctly else the databases can't be found. The installation step just adds a platform-specific binary file, but in this case you could include your own wrapper script if desired.